test/core/networking/websockets/session_test.dart

```diff
@@ -0,0 +1,408 @@
+    client.onMessage = (ClientSession sess, dynamic msg) async {
+      sess.raise(TestException(msg));
+    };
+    client.onError = (ClientSession sess, Object ex) {
+      exceptionClient.set(ex);
+    };
+
+    // Wait for handlers to be called
+    await exceptionClient.awaitValue(Duration(seconds: 5), raiseOnTimeout: true);
+
+    // Check exception
+    exceptionClient.expect(
+      equals(
+        isA<TestException>().having(
+          (TestException e) => e.message,
+          'message',
+          equals('ANYTHING-2'),
+        ),
+      ),
+    );
+  });
+  test('OnMessage return is send', () async {
+    final VHook<String?> msgReceivedFromServer = VHook<String?>(null);
+    final VHook<String?> msgReceivedFromClient = VHook<String?>(null);
+
+    //* Test server
+    final StreamSubscription<HttpRequest> sub = server!.listen((HttpRequest request) {
+      final ServerSession sess = ServerSession(request);
+      sess.onMessage = (ServerSession sess, dynamic msg) async {
+        return 'ANYTHING-RETURN-1';
+      };
+    });
+
+    // Connect to WebServer and open ServerSession
+    ClientSession client = ClientSession(url.toString());
+    //? This is not a race condition, because of how the dart eventloop is implemented
+    //? The internal 'WebSocket.connect' is scheduled and executed when this strain waits (await)
+    client.onMessage = (ClientSession sess, dynamic msg) async {
+      msgReceivedFromServer.set(msg);
+    };
+
+    // Send message to trigger onMessage return
+    client.send('ANYTHING');
+
+    // Wait for msgs to be received
+    await msgReceivedFromServer.awaitValue(Duration(seconds: 5), raiseOnTimeout: true);
+
+    // Check received data
+    msgReceivedFromServer.expect(equals('ANYTHING-RETURN-1'));
+
+    //* Test client
+    sub.onData((HttpRequest request) {
+      final ServerSession sess = ServerSession(request);
+      sess.onMessage = (ServerSession sess, dynamic msg) async {
+        msgReceivedFromClient.set(msg);
+      };
+
+      // Send message to trigger onMessage return
+      sess.send('ANYTHING');
+    });
+    await client.close();
+
+    // Connect to WebServer and open ServerSession
+    client = ClientSession(url.toString());
+    //? This is not a race condition, because of how the dart eventloop is implemented
+    //? The internal 'WebSocket.connect' is scheduled and executed when this strain waits (await)
+    client.onMessage = (ClientSession sess, dynamic msg) async {
+      return 'ANYTHING-RETURN-2';
+    };
+
+    // Wait for msgs to be received
+    await msgReceivedFromClient.awaitValue(Duration(seconds: 5), raiseOnTimeout: true);
+
+    // Check received data
+    msgReceivedFromClient.expect(equals('ANYTHING-RETURN-2'));
+
+    // Teardown
+    client.close();
+  });
+}
+
+class TestException implements Exception {
+  final String message;
+
+  const TestException(this.message);
+
+  @override
+  String toString() {
+    return message;
+  }
+}
```

Comment:
- Docs 
- Private or custom file
- The same class is in server_test. You should really move it to it's own file and reuse the code

Reply:
Docs - yes
Private - why?!
Custom file - maybe

Reply:
Did you read the wikipedia about DRY. That's not DRY, because you repeat yourself here